design-proposals/backend-storage-block-migration.md

```diff
@@ -0,0 +1,62 @@
+# Overview
+This proposal is about supporting the migration of all VMs that use a persistent feature, such as persistent vTPM.
+
+To make this document as simple as possible, let's consider that storage in Kubernetes can either be filesystem (FS) or
+block mode, and either ReadWriteOnce (RWO) or ReadWriteMany (RWX) access mode, with storage classes supporting any
+number of combinations of both.
+
+As of KubeVirt 1.3, Here is how each of the 4 combinations is supported by the backend-storage for persistent VM state:
+- RWO FS: supported by backend-storage but makes VMs non-migratable
+- RWX FS: preferred combination, supported by backend-storage and compatible with VM migrations
+- RWO block: not supported yet, but PRs exist to implement it in a non-migratable way 
+- RWX block: same as above, with migratability functional but considered unsafe (as it concurrently mounts an ext3 partition)
```

Comment:
This is because, unlike with VM disks, we have 2 components concurrently accessing the PVC (source and target virt-launchers).
They both maintain the filesystem metadata for the mounted partition, which will lead to conflicts. I can add more verbiage to explain that in the doc.

Reply:
> I can add more verbiage to explain that in the doc.

Please do. In particularly, please refer to my question

> Does libvirt have the same problem? If the only shared storage a user has is block, can they use vTPM+migration?

Reply:
Currently the TPM implementation uses multiple files under the hood.

This means one of two things must be true. Either some component in the stack must manage a filesystem on the block device or the TPM implementation must be rewritten from the ground up. The latter doesn't seem like a fair request to the underlying projects. At least not in a timeline that suits our needs.

The filesystem being used is ext3, so it can either be mounted on the source or destination only. These mounts are mutually exclusive with each other regardless of whether or not the PVC is using RWX mode. This is simply a limitation of ext3 (or most filesystems for that matter).

Given these constraints, KubeVirt is the best component to manage this problem because it can manage PVCs where libvirt can't--because libvirt runs inside a container.

Reply:
@stu-gott do I understand that your the answer to my question is affirmative? libvirt has the same problem and cannot cannot use shared block+vTPM+migration?

Have we asked them to support this combination?
Do they not mind?

Reply:
@dankenigsberg I answered in another comment. The answer is, to the best of my knowledge, no.
swtpm, qemu and/or libvirt don't support saving tpm state on block storage and there is no plan to support this.
If we want to support this scenario in KubeVirt we will need to provide the management around this.